File: Polygon.js
var vertices = 6;
var noOfPolygons = 50;
var geneSize = 4 + (vertices * 2);
var dnaLength = noOfPolygons * geneSize;

function Polygon(dna, fitness){

	if(dna && fitness){
		this.dna = dna;
		this.fitness = fitness;
	} else if(dna){
		this.dna = dna;
	} else{
		this.dna = new DNA();
	}

	this.show = function(){
		for(var i = 0; i < dnaLength; i+=geneSize){
			beginShape();
				for(var j = i+4; j < i+geneSize; j+=2){
					vertex((this.dna.genes[j] * 200) >> 0, (this.dna.genes[j+1] * 200) >> 0);
				}
				noStroke();
				fill((this.dna.genes[i] * 255) >> 0, (this.dna.genes[i+1] * 255) >> 0, (this.dna.genes[i+2] * 255) >> 0, (this.dna.genes[i+3] * 255) >> 0);
			endShape();
		}
	}

	this.calculateFitness = function(){
		var diff = 0;
		workingCtx.clearRect(0, 0, 200, 200);
		this.show();
		loadPixels();
		var diff = 0;
		for(var i = 0; i < 200*200*4; i++){
			var dp = data.data[i] - pixels[i];
			diff += dp * dp;
		}
		this.fitness = 1 - diff / (200 * 200 * 4 * 255 * 255);
	}

	/*this.breed = function(parent){
		var child = this.dna.breed(parent.dna);
		return (new Polygon(child));
	}*/
	this.breed = function(){
		var cloneParent = this.clone();
		var child = this.dna.breed();
		var childPolygon = new Polygon(child);
		childPolygon.calculateFitness();
		if(childPolygon.fitness > this.fitness){
			return childPolygon;
		} else {
			return cloneParent;
		}
	}
	this.clone = function(){
		var cloneDNA = this.dna;
		var cloneFitness = this.fitness;
		return (new Polygon(cloneDNA, cloneFitness));
	}

	this.drawToASpecificContext = function(context){
		context.clearRect(0, 0, 200, 200);
		for(var i = 0; i < dnaLength; i+=geneSize){
			context.beginPath();
			context.moveTo((this.dna.genes[i + 4] * 200) >> 0, (this.dna.genes[i + 5] * 200) >> 0);

			for(var j = i+6; j < i+geneSize; j+=2){
				context.lineTo((this.dna.genes[j] * 200) >> 0, (this.dna.genes[j+1] * 200) >> 0);
			}
			context.closePath();	

			var styleString = 'rgba(' + 
				((this.dna.genes[i] * 255) >> 0) + ',' + 
		        ((this.dna.genes[i+1] * 255) >> 0) + ',' + 
		        ((this.dna.genes[i+2] * 255) >> 0) + ',' +
		        (this.dna.genes[i+3]) + ')';
		    context.fillStyle = styleString;
		    context.fill();
		}
	}
}

File: python/Polygon.py
import DNA
import settings
from PIL import Image, ImageDraw


class Polygon(object):

	def __init__(self, dna=None, fitness=None):
		if(dna and fitness):
			self.dna = dna;
			self.fitness = fitness;
		elif(dna):
			self.dna = dna;
		else:
			self.dna = DNA.DNA();

	def createImageFromGenes(self):
		img = []
		draw = []
		temp = self.dna.genes
		for i in range(0, settings.dnaLength, settings.geneSize):
			tempIMG = Image.new('RGBA', (200, 200));
			img.append(tempIMG);
			tempDRAW = ImageDraw.Draw(tempIMG);
			draw.append(tempDRAW);
			tempDRAW.polygon([(int(temp[i+4]*200), int(temp[i+5]*200)),(int(temp[i+6]*200), int(temp[i+7]*200)),
						(int(temp[i+8]*200), int(temp[i+9]*200))], 
						fill = (int(temp[i]*255),int(temp[i+1]*255),int(temp[i+2]*255),int(temp[i+3]*255)));

		settings.finalIMG = Image.new('RGBA', (200, 200));
		for i in range(settings.noOfPolygons):
			settings.finalIMG = Image.alpha_composite(settings.finalIMG, img[i])
		return settings.finalIMG.load()


	def calculateFitness(self):
		diff = 0;
		dp = 0;
		pixels = self.pix_to_seq(self.createImageFromGenes(), 0)
		for i in range(200*200*4):
			dp = settings.sourcePIX[i] - pixels[i]
			diff += dp * dp
		self.fitness = 1 - diff / (200 * 200 * 4 * 255 * 255)

	def breed(self):
		cloneParent = Polygon(self.dna, self.fitness);
		child = self.dna.breed();
		childPolygon = Polygon(child);
		childPolygon.calculateFitness();
		if(childPolygon.fitness > self.fitness):
			return childPolygon;
		else:
			return cloneParent;

	def saveImage(self):
		result = Image.new('RGBA', (200, 200));
		result.putdata(data=self.pix_to_seq(settings.finalIMG.load(), 1));
		result.save("result.png");

	def pix_to_seq(self, pix, opt):
		seq1 = []
		seq2 = []
		for i in range(200):
			for j in range(200):
				seq2.append(pix[i,j])
				for k in range(4):
					seq1.append(pix[i,j][k])
		if(opt == 0):
			return seq1
		else:
			return seq2